### DISCUSSION TITLE
SSR/next-urql questions

### DISCUSSION THREAD

**Discussion by saevarb** [urql-graphql/urql#2436] [Q&A] [answered]
Hello.

For various reasons, I have been using the `next-urql` and `ssrExchange` for a while in one of my projects without SSR actually ever working. It wasn't a hard requirement, so I opted to let it be, but I now finally have some time to look into it.

For background, I am doing exclusively SSR and no static rendering of any kind, and the graphql API is served from the same host as the web app, I am wrapping my entire `_app` in `withUrqlClient`, and my `_app` component also has `getInitialProps`.

From reading the docs for `next-urql`, my immediate impression is that there are some caveats if you want to wrapp your entire `_app`, but that it's possible, and if I do, it should "just work". There is a link to an example, but it's dead and I can't find it [here](https://github.com/FormidableLabs/urql/tree/main/examples) or anywhere else.

But then when I read further, there is this section:
> In client-side SPAs using urql, you typically configure the Client yourself and pass it as the value prop to urql's context Provider. withUrqlClient handles setting all of this up for you under the hood. By default, you'll be opted into server-side Suspense and have the necessary exchanges set up for you, including the [ssrExchange](https://formidable.com/open-source/urql/docs/api/#ssrexchange-exchange-factory).

This section now seems to be telling me that having `Suspense` support is required. I understand the requirement, but as someone reading the documentation who doesn't have a full understanding of all the underlying mechanisms, what I have read up until now has led me to believe that it would be possible to use the SSR functionality without it(which, if I understand correctly, is still possible, but it requires you to first prime your cache manually in `getInitialProps`?).

I also think it's relatively recent that creating a nextjs project with `create-next-app` would default to using react 18. My current project is only 3 months old and is still using react 17, and trying to use `next-urql` with it gives a `ReactDOMServer does not yet support Suspense`.

The good news is that if I just upgrade to react 18 and ensure `neverSuspend: false`, it Just Works, which is great. 

I feel like perhaps the docs could be improved a bit, and I'd be open to making a pull request, but I would have to figure out what the correct thing is to do. I.e. for recent versions of next, you can just upgrade to react 18, but I think for others you might need to enable experimental features, and still others it might not work at all.

Also, I'm interested in trying to write something similar to apollo's `SchemaLink` to execute SSR queries directly on the server. I was just going to dive in and read your docs and read existing exchanges, but if there is any particular prior work you think I should be looking at, I would appreciate any pointers.

**Comment by JoviDeCroock** [comment, marked as answer]
The Suspense support is a requirement under the hood for [react-ssr-prepass](https://github.com/FormidableLabs/react-ssr-prepass) you find an example [here](https://github.com/FormidableLabs/urql/blob/main/examples/with-next/pages/index.js) so basically if you opt to use `getInitialProps` over any of the getXProps exported functions in Next we will run `react-ssr-prepass` for you which will wait at every query for a given subtree and resolve the data. Then we cache this data in the `ssrExchange` and next will run `renderToString` basically for this automatic way you don't have to worry.

The schema link exchange can be found [here](https://github.com/FormidableLabs/urql/tree/main/exchanges/execute)

**Comment by maltesa** [comment]
> For various reasons, I have been using the next-urql and ssrExchange for a while in one of my projects without SSR actually ever working. It wasn't a hard requirement, so I opted to let it be, but I now finally have some time to look into it.

This is exactly how I feel. I put a lot of effort into figuring out how to use getInitialProps next-urql and got it working (using executeExchange during build etc.), but at some point it broke again (cache was generated during built and hydrated, but no refetch of stale data on intital page load, tried with and without staleWhileRevalidate. Also, my local graphcache resolvers were not working anymore). I'm using the latest package versions and latest stable nextjs with react 18.

The documentation is not clear ssr works when using graphcache. Hence, I also tried using the “simple” cacheExchange. But without success. I would be very happy If the documentation would provide a more clear and complete example on how to make it work with `getInitialProps`. I think this is the main use-case for many nextjs users, especially with incremental static regeneration it would be awesome!

## How I tried to use it
### Helper:
```ts
export const url = '/api/graphql'
export const baseExchanges = [devtoolsExchange, dedupExchange, graphCacheExchange]

async function getSSRClientAndCache() {
  const ssrCache = ssrExchange({ isClient: false })
  const schemaExchange = await initSchemaExchange()
  const client = initUrqlClient(
    { url, exchanges: [...baseExchanges, ssrCache, schemaExchange], suspense: true },
    false
  )

  return { client, ssrCache }
}

type TQueryList = { query: DocumentNode; variables?: Record<string, any> }[]

export async function populateGqlCache(queries: TQueryList) {
  const { client, ssrCache } = await getSSRClientAndCache()
  if (!client) return

  // Call queries to popluate cache
  for (const queryAndVars of queries) {
    const { query, variables } = queryAndVars
    const { error } = await client.query(query, variables).toPromise()
    if (error) throw error
  }

  return ssrCache.extractData()
}

export const withGraphQlClient = withUrqlClient(
  (ssrCache) => {
    return {
      url,
      exchanges: [...baseExchanges, ssrCache, fetchExchange],
      requestPolicy: 'cache-and-network',
    }
  },
  { ssr: false, staleWhileRevalidate: true }
)

```

### Next Page
```tsx
export const getStaticProps: GetStaticProps = async () => {
  const urqlState = await populateGqlCache([{ query: CommunityPostsDocument }])

  return {
    props: { urqlState },
    revalidate: 60 * 5, // 5 minutes
  }
}

function Community() {
  const [{ data }] = useCommunityPostsQuery()

  if (!data?.communityPosts) return <>Loading...</>

  return (
   <>
     {// Render list of posts}
    </>
  )
}

export default withGraphQlClient(Community)
```

## How I dream it would work
- tell `withUrqlClient` which exchanges to use server-side (executeExchange instead of fetch)
- wrap app component withUrqlClient
- all pages with getInitialProps or getServerSideProps build the cache automatically

Is that even possible? I thought it might, with suspense starting to work server-side?
